README
Switch to http://github.com/DrupalVietNam/document


examples/CHANGELOG.txt
# $Id: CHANGELOG.txt,v 1.1.2.4 2010/09/26 13:16:24 rfay Exp $




index.php
<?php
// $Id: index.php 2 2010-12-28 08:18:19Z drupalvietnam $

/**
 * @mainpage Tài liệu cho Drupal Developer -- DrupalVietnam.org
 * Chào mừng các bạn đến với trang tài liệu dành cho người phát triển Drupal.
 * Cung cấp bởi @link http://drupalvietnam.org/ Drupal Vietnam Community @endlink
 *
 * - Drupal Core
 *  - @link drupal_coding_standards Chuẩn lập trình Drupal @endlink
 *  - @link drupal_database_usage Sử dụng Database API @endlink
 *  - @link drupal_form_api_usage Sử dụng Form API @endlink
 *  - @link drupal_form_api_reference Tham khảo Form API @endlink
 *  - @link http://drupal.org/project/examples Các module mẫu @endlink các tính
 *    năng được Drupal Core cung cấp.
 *  - Drupal và Javascript.
 *  - @link http://drupal.org/project/examples Các module mẫu @endlink
 *    - @link AHAH
 *
 * - Module mở rộng
 *  - @link http://drupal.org/project/ctools CTools @endlink
 *    - Giới thiệu CTools
 *  - @link http://drupal.org/project/views Views @endlink
 *    - Giới thiệu Views
 *    - ...
 *  - @link http://drupal.org/project/cck CCK @endlink
 *    - Giới thiệu CCK
 *    - ...
 * - Xây dựng giao diện cho Drupal
 *  - Giới thiệu PHPTemplate và cơ chế quá tải cho theme.
 *  - ...
 *  - Giới thiệu module devel_themer.
 */


drupal.database-api.php
<?php
// $Id: drupal.database-api.php 1 2010-12-28 04:38:52Z drupalvietnam $

/**
 * @defgroup drupal_database_usage Sử dụng Database API.
 * @{
 * Từ Drupal 7
 *
 * Thí dụ db_insert():
 * @code
 *  // INSERT INTO {dbtng_example} (name, surname) VALUES('John, 'Doe')
 *  db_insert('dbtng_example')
 *    ->fields(array('name' => 'John', 'surname' => 'Doe'))
 *    ->execute();
 * @endcode
 *
 * Thí dụ db_update():
 * @code
 *  // UPDATE {dbtng_example} SET name = 'Jane' WHERE name = 'John'
 *  db_update('dbtng_example')
 *    ->fields(array('name' => 'Jane'))
 *    ->condition('name', 'John')
 *    ->execute();
 * @endcode
 *
 * Thí dụ db_delete():
 * @code
 *  // DELETE FROM {dbtng_example} WHERE name = 'Jane'
 *  db_delete('dbtng_example')
 *    ->condition('name', 'Jane')
 *    ->execute();
 * @endcode
 */
/**
 * @} End of "defgroup drupal_database_usage"
 */


drupal.form-api-usage.php
<?php
// $Id: drupal.form-api-usage.php 1 2010-12-28 04:38:52Z drupalvietnam $

/**
 * @defgroup drupal_form_api_usage Sử dụng Drupal Form API
 * @{
 * Sử dụng Drupal, bạn nên quên đi cách viết các thẻ HTML để tạo ra form, nhưng
 * hãy sử dụng Form API. Form API hỗ trợ người phát triển tạo ra form nhanh, gọn,
 * bảo mật, với khả năng mở rộng cực kỳ mạnh.
 *
 * Với cơ chế hook quen thuộc của Drupal, Form API trở nên một vũ khí lợi hại là
 * một phần không thể không nhắc tới khi nói đến Drupal. Thí dụ:
 *  - Module WYSIWYG alter vào các form, thay đổi textarea thành các rich-text-editor. 
 *  - Module Logintobogan alter vào form login, mở rộng khả năng login bằng email
 *    cho người dùng.
 *  - Captcach alter vào form để thể captcha field.
 *  - ...
 *
 * Để áp dụng được rất nhiều tính năng sẵn có, dựa trên cơ chế hook vào form,
 * thì các form của bạn tạo ra, không cách nào khác, phải sử dụng Form API. Tuy
 * nhiên, việc sử dụng Form API rất đơn giản.
 *
 * Về cơ bản, Form API chia luồng xử lý form của bạn ra làm 3 phần: hàm tạo cấu
 * trúc form, validate form và phần xử lý dữ liệu được đệ trình. Để dễ hiểu, ở 
 * đây chúng ta làm một thí dụ: Tạo một form chỉ chứa một text field, để người 
 * dùng nhập email address, và một nút submit (có thể xem như một subcription 
 * form).
 *
 * Như đã nói ở trên, bước một chúng ta tạo ra hàm để định nghĩa form. Chúng ta
 * có thể chọn một tên hàm bất kỳ, bắt đầu bằng tên của module mà chúng ta đang
 * làm việc (giả sử là www).
 *
 * @code
 *  function www_subscription_form(&$form_state) {
 *    // Text field để người dùng nhập vào địa chỉ email.
 *    $form['email'] = array(
 *      '#type' => 'textfield',
 *      '#title' => 'Địa chỉ email',
 *      '#required' => TRUE,
 *    );
 *    
 *    // Nút submit
 *    $form['submit'] = array(
 *      '#type' => 'submit',
 *      '#value' => 'Gửi đi',
 *    );
 *
 *    return $form;
 *  }
 * @endcode
 *
 * Chúng ta có thể nhận ra rằng, cấu trúc một form làm một mảng, với các phần tử
 * mảng chính là các field. Chúng ta có thể tham khảo thêm ở
 * @link http://api.drupal.org/api/drupal/developer--topics--forms_api_reference.html/6 Forms API Reference @endlink
 * để biết những phần tử form được Drupal hỗ trợ sẵn trong core.
 *
 * Tiếp đến phần thứ hai, chúng ta xây dựng thêm một hàm để validate nội dung 
 * người dùng truyền vào form. Mặc định tên hàm validate sẽ là tên của hàm đã 
 * định nghĩa ra form cộng với hậu tố _validate, bước một chúng ta đã chọn
 * www_subscription_form, vậy tên hàm validate của chúng ta sẽ là
 * www_subscription_form_validate:
 *
 * @code
 *  function www_subscription_form_validate(&$form, &$form_state) {
 *    // Ở đây, để đơn giản, chúng ta chỉ validate nội dung người dùng
 *    // nhập vào text field là một email hợp lệ
 *    if (!valid_email_address($form_state['values']['email'])) {
 *      // Báo cho Drupal biết là field email có lỗi
 *      // và hiển thị thông báo lỗi ra màn hình người dùng
 *      form_set_error('email', 'Bạn đã nhập email không hợp lệ.');
 *    }
 *  }
 * @endcode
 *
 * Chúng ta có thể yên tâm, nếu dữ liệu được gửi đến không pass được hàm validate
 * (hàm form_set_error() được gọi), thì phần xử lý form sẽ không được gọi.
 * $form và biến định nghĩa cấu trúc của form, $form_state là biến chứa tình
 * trạng hiện tại của form, bao gồm dữ liệu nhận được từ người dùng.
 *
 * Phần 3, là phần chúng ta xử lý nội dung, đã hợp lệ, mà người dùng gửi tới.
 * Chúng ta cũng phải cần định nghĩa thêm một hàm nữa, mặc định tên hàm sẽ là
 * tên của hàm định nghĩa ra form cộng với hâu tố _submit:
 *
 * @code
 *  function www_subscription_form_submit(&$form, &$form_state) {
 *    // Lưu email vào database
 *    // ...
 *    // Trình bày thông điệp - xử lý thành công - ra màn hình người dùng
 *    drupal_set_message(sprintf(
 *      "Cảm ơn... Chúng tôi đã lưu email <strong>%s</strong> vào CSDL.",
 *      $form_state['values']['email']
 *    ));
 *  }
 * @endcode
 *
 * @TODO: 1. Multistep form
 * @TODO: 2. Định nghĩa thêm form-field.
 * @TODO: 3. Tuỳ biến giao diện form.
 * @TODO: 4. Mở rộng, tuỳ biến các form hiện có.
 * @TODO: 5. Thí dụ tạo form cho phép upload hỉnh ảnh.
 */
/**
 * @} End of "defgroup drupal_form_api_usage"
 */


themes/drupalvietnam/tpl/page.tpl.php
<?php
// $Id: page.tpl.php,v 1.43 2010/01/30 07:59:25 dries Exp $

/**
 * @file
 * Default theme implementation to display a single Drupal page.
 *
 * Available variables:
 *
 * General utility variables:
 * - $base_path: The base URL path of the Drupal installation. At the very
 *   least, this will always default to /.
 * - $directory: The directory the template is located in, e.g. modules/system
 *   or themes/garland.
 * - $is_front: TRUE if the current page is the front page.
 * - $logged_in: TRUE if the user is registered and signed in.
 * - $is_admin: TRUE if the user has permission to access administration pages.
 *
 * Site identity:
 * - $front_page: The URL of the front page. Use this instead of $base_path,
 *   when linking to the front page. This includes the language domain or
 *   prefix.
 * - $logo: The path to the logo image, as defined in theme configuration.
 * - $site_name: The name of the site, empty when display has been disabled
 *   in theme settings.
 * - $site_slogan: The slogan of the site, empty when display has been disabled
 *   in theme settings.
 *
 * Navigation:
 * - $main_menu (array): An array containing the Main menu links for the
 *   site, if they have been configured.
 * - $secondary_menu (array): An array containing the Secondary menu links for
 *   the site, if they have been configured.
 * - $breadcrumb: The breadcrumb trail for the current page.
 *
 * Page content (in order of occurrence in the default page.tpl.php):
 * - $title_prefix (array): An array containing additional output populated by
 *   modules, intended to be displayed in front of the main title tag that
 *   appears in the template.
 * - $title: The page title, for use in the actual HTML content.
 * - $title_suffix (array): An array containing additional output populated by
 *   modules, intended to be displayed after the main title tag that appears in
 *   the template.
 * - $messages: HTML for status and error messages. Should be displayed
 *   prominently.
 * - $tabs (array): Tabs linking to any sub-pages beneath the current page
 *   (e.g., the view and edit tabs when displaying a node).
 * - $action_links (array): Actions local to the page, such as 'Add menu' on the
 *   menu administration interface.
 * - $feed_icons: A string of all feed icons for the current page.
 * - $node: The node object, if there is an automatically-loaded node
 *   associated with the page, and the node ID is the second argument
 *   in the page's path (e.g. node/12345 and node/12345/revisions, but not
 *   comment/reply/12345).
 *
 * Regions:
 * - $page['help']: Dynamic help text, mostly for admin pages.
 * - $page['highlight']: Items for the highlighted content region.
 * - $page['content']: The main content of the current page.
 * - $page['sidebar_first']: Items for the first sidebar.
 * - $page['sidebar_second']: Items for the second sidebar.
 * - $page['header']: Items for the header region.
 * - $page['footer']: Items for the footer region.
 *
 * @see template_preprocess()
 * @see template_preprocess_page()
 * @see template_process()
 */
?>

  <div id="page-wrapper"><div id="page">

    <div id="header"><div class="section container clearfix">
    
      <?php if ($primary_nav): ?>
        <div id="navigation" class="clearfix">
          <?php print $primary_nav; ?>
        </div> <!-- /#navigation -->
      <?php endif; ?>
      
      <div id="header-content">
        <div id="header-left" class="span-12">
        <?php if ($site_name || $site_slogan): ?>
          <div id="name-and-slogan">
            <?php if ($site_name): ?>
                <h1 id="site-name">
                  <a href="<?php print $front_page; ?>" title="<?php print t('Home'); ?>" rel="home"><?php print $site_name; ?></a>
                </h1>
            <?php endif; ?>
  
            <?php if ($site_slogan && drupal_is_front_page()): ?>
              <h2 id="site-slogan"><?php print $site_slogan; ?></h2>
              <?php print render($page['header_front']); ?>
            <?php endif; ?>
          </div> <!-- /#name-and-slogan -->
        <?php endif; ?>
        </div>
        <div id="header-right" class="prepend-2 span-10 last">
        <?php print render($page['search']); ?>
        </div>
      </div>
      

      <?php print render($page['header']); ?>

    </div></div> <!-- /.section, /#header -->


    <div id="main-wrapper"><div id="main" class="clearfix container showgrid">
    <?php if ($breadcrumb): ?>
      <div id="breadcrumb"><?php print $breadcrumb; ?></div>
    <?php endif; ?>

    <?php print $messages; ?>
    
      <div id="content" class="column"><div class="section">
        <a id="main-content"></a>
        <?php print render($title_prefix); ?>
        <?php if ($title): ?><h1 class="title" id="page-title"><?php print $title; ?></h1><?php endif; ?>
        <?php print render($title_suffix); ?>
        <?php if ($tabs): ?><div class="tabs"><?php print render($tabs); ?></div><?php endif; ?>
        <?php print render($page['help']); ?>
        <?php if ($action_links): ?><ul class="action-links"><?php print render($action_links); ?></ul><?php endif; ?>
        <?php print render($page['content']); ?>
        <?php print $feed_icons; ?>
      </div></div> <!-- /.section, /#content -->

    </div></div> <!-- /#main, /#main-wrapper -->

    <div id="footer"><div class="section container">
      <?php print render($page['footer']); ?>
    </div></div> <!-- /.section, /#footer -->

  </div></div> <!-- /#page, /#page-wrapper -->


drupal.coding-standards.php
<?php
// $Id: drupal.coding-standards.php 1 2010-12-28 04:38:52Z drupalvietnam $

/**
 * @defgroup drupal_coding_standards Chuẩn viết mã Drupal
 * @{
 *
 * - Mỗi người có cách viết code khác nhau, không thể nói cách viết của ai là
 *   đúng, là sai, là tốt hơn một cách khách quan được.
 *   Tuy nhiên, để người khác có dễ dàng hiểu code của mình, và ngược lại, khi
 *   làm việc với mã nguồn Drupal, bạn cần tuân thủ Chuẩn viết mã của Drupal --
 *   những quy ước viết code được cộng đồng đề xuất. Bạn có thể tham khảo chi tiết
 *   tại @link http://drupal.org/coding-standards Drupal.org @endlink.
 *
 * - Nguyên tắc không hack core.
 *   
 *   Drupal cung cấp cơ chế rất linh động cho người phát triển cuối -- bạn có
 *   thể thay thế các hành vi mặc định của Drupal thành hành vi tuỳ biến của bạn
 *   mà không cần chỉnh sửa trực tiếp vào nhân của Drupal.
 *   
 *   Trường hợp bạn buộc phải hack core (hỗ trợ CDN, hiphop, ...), bạn phải tự
 *   giải quyết những vấn đề phát sinh, những thay đổi của bạn sẽ mất nếu bạn
 *   cập nhật mã nguồn mới nhất, ...
 *
 *   Trường hợp bạn thấy Drupal chưa thật sự uyển chuyển ở phần nào đó, bạn có
 *   thể tạo issue mới để yêu cầu tính năng, những người có kinh nghiệm sẽ đề
 *   suất cách giải quyết cho bạn, hoặc họ sẽ tạo patch cho tính năng mà bạn đề
 *   suất.
 *
 * - Tổ chức file trong Drupal
 *   Drupal có quy ước tổ chức mã nguồn riêng của nó, nắm bắt được quy ước này,
 *   chúng ta có thể xây dựng được modules/themes/profiles có thể sử dụng ở mọi
 *   website chạy Drupal (thậm chí, nhiều site Drupal chạy cùng trên một mã 
 *   nguồn).
 *   
 *    Xem cấu trúc tập tin của Drupal, chúng ta thấy các các phần sau:
 *    
 *    - (01) includes
 *        Các thư viện cơ bản cho Drupal, đa phần chúng swappable, nghĩa là chúng
 *        ta có thể buộc Drupal sử dụng thư viện thay thế khác, mà không cần
 *        chỉnh sửa gì lên thư viện hiện tại.
 *        Thí dụ, nếu muốn sử dụng thư viện cache.inc mặc định của Drupal, cache
 *        trên database, thành thư viện cache.memcache.inc, thì ở tập tin cấu
 *        hình của Drupal site (thường là sites/default/settings.php), chúng ta
 *        thêm vào đó dòng code sau
 *        
 *          $conf['cache_inc'] = '/path/to/cache.memcache.inc';
 *        
 *        Cho nên, đối với thư mục này của Drupal, chúng ta không nên chỉnh sửa gì.
 *    - (02) misc
 *        Là thư mục chứa những hình ảnh, css và thư viện javascript được sử
 *        dụng trong hệ thống.
 *    - (03) modules
 *    - (04) modules/*
 *    - (05) profiles
 *    - (06) scripts
 *    - (07) sites
 *    - (08) sites/all
 *    - (09) sites/default
 *    - (10) sites/default/default.settings.php
 *    - (11) themes
 *    - (12) .htaccess
 *    - (13) *.txt
 *    - (14) robots.txt
 *
 * - Document mã nguồn
 *   
 *   Để mã nguồn dễ hiểu, một việc tất yếu, bạn phải chú thích, viết tài liệu
 *   cản thận, chi tiết cho các class, interface, method, function, ... của mình.
 *   Nếu phần documetn của bạn đúng theo cú pháp được 
 *   @link http://drupal.org/project/api API module @endlink hỗ trợ, bạn không 
 *   cần phải tốn thêm thời gian để soạn thảo lại tài liệu. 
 *   @link http://drupal.org/project/api API module @endlink hỗ trợ trự động phát
 *   sinh trang tham khảo (giống như
 *   @link http://api.drupal.org api.drupal.org @endlink) từ chính mã nguồn bạn 
 *   đã viết.
 *   @link http://drupal.org/node/1354 Tham khảo các cú pháp được Doxygen hỗ trợ (Doxygen formatting conventions) @endlink.
 *
 */
/**
 * @} End of "defgroup drupal_coding_standards"
 */
